File: kotlin/src/main/kotlin/adventofcode2022/Day07.kt
package adventofcode2022

import adventofcode2022.util.runWithStopwatch
import java.io.File

fun main() {
    runWithStopwatch { part1() }
    runWithStopwatch { part2() }
}

private fun part2() {
    val commands = File("./input/day07.txt").readLines()
    val directories = buildDirectoryStructure(commands)
    val rootMemory = directories.values.toList()[0]
    val totalMemory = 70000000
    val spaceNeeded = 30000000
    val deleteSize = spaceNeeded - (totalMemory - rootMemory)
    val directoryToDelete = directories
            .map { it.value }
            .filter { it >= deleteSize }
            .minOf { it }
    println("Director to delete has memory: $directoryToDelete")
}

private fun part1() {
    val commands = File("./input/day07.txt").readLines()
    val directories = buildDirectoryStructure(commands)
    val sortedDescending = directories.map { it.value }.toList().sortedDescending()
    val totalValue = sortedDescending.filter { it <= 100000 }.sum()
    println("Total value: $totalValue")

}

private fun buildDirectoryStructure(commands: List<String>): MutableMap<String, Long> {
    val directories = mutableMapOf<String, Long>()
    var path = ""
    for (command in commands) {
        if ("$ cd " in command) {
            val input = command.split("$ cd ")
            when (input[1]) {
                ".." -> {
                    path = path.substring(0, path.length - path.split("-").last().length - 1)
                }

                else -> {
                    path += "-" + input[1]
                    directories[path] = 0
                }
            }
            continue
        }
        val input = command.split(" ")
        val memory = input[0].toLongOrNull()
        if (memory != null) {
            for (element in directories) {
                if (element.key in path) {
                    directories[element.key] = directories[element.key]!! + memory
                }
            }
        }
    }
    return directories
}

File: python/Day07.py

from utils.Stopwatch import runWithStopwatch


def part1():
    commands = open("../input/day07.txt", "r").readlines()
    structure = create_directory_structure(commands)
    total_value = 0
    for value in structure.values():
        if value >= 100000:
            total_value += value
    print("Total memory of big directories are: " + str(total_value))


def create_directory_structure(commands):
    structure = dict()
    directory = ".."
    for command in commands:
        command = command.strip()
        if "$ cd " in command:
            directory = command.split("$ cd ")[1]
            if directory == "..":
                continue
            else:
                structure[directory] = 0
        command = command.split(" ")
        if command[0].isdigit():
            structure[directory] += int(command[0])
    return structure


def part2():
    pass


if __name__ == '__main__':
    runWithStopwatch(part1)
    runWithStopwatch(part2)
